homeassistant/components/synology_dsm/__init__.py

```diff
@@ -179,6 +194,22 @@ def _async_migrator(entity_entry: entity_registry.RegistryEntry):
 
     await entity_registry.async_migrate_entries(hass, entry.entry_id, _async_migrator)
 
+    # migrate device indetifiers
+    dev_reg = await get_dev_reg(hass)
+    devices: list[DeviceEntry] = device_registry.async_entries_for_config_entry(
+        dev_reg, entry.entry_id
+    )
+    for device in devices:
+        old_identifier = list(next(iter(device.identifiers)))
+        if len(old_identifier) > 2:
+            new_identifier = {
+                (old_identifier.pop(0), "_".join([str(x) for x in old_identifier]))
+            }
+            _LOGGER.debug(
+                "migrate identifier '%s' to '%s'", device.identifiers, new_identifier
+            )
+            dev_reg.async_update_device(device.id, new_identifiers=new_identifier)  # type: ignore[arg-type]
```

Comment:
Withoutu that `ignore` mypy throws the following - but I do not have any glue, how to solve that :confused: 
```
homeassistant/components/synology_dsm/__init__.py:211: error: Argument "new_identifiers" to "async_update_device" of "DeviceRegistry" has incompatible type "Set[Tuple[str, str]]"; expected "Union[Set[Tuple[str, ...]], UndefinedType]"  [arg-type]
```

Reply:
That's because of invariance: https://mypy.readthedocs.io/en/latest/common_issues.html#invariance-vs-covariance

Type from collections.abc should be used to support tuple[str, str]. MutableSet or AbstractSet.

Or you can define the same type of tuple[str, ...].

cc @scop

Reply:
I'm sorry, but to be honest, I did not understand anything from that comment :see_no_evil:

Reply:
tl/dr:

set[tuple[str, str]] is not compatible with set[tuple[str, ...]]. It needs exact match.

But AbstractSet[tuple[str, str]] is compatible with AbstractSet[tuple[str, ...]].

Reply:
OK ... should I now simple add definition `new_identifier: set[tuple[str, ...]]` or should this be solved in `DeviceRegistry.async_update_device()` :thinking:

Reply:
For this PR you can use the first option.